R/vcov.R

```diff
@@ -23,11 +22,107 @@
 #'
 #' @method vcov synthdid_estimate
 #' @export
-vcov.synthdid_estimate = function(
-  object,
+vcov.synthdid_estimate = function(object,
   method = c("bootstrap", "jackknife", "placebo"),
-  weights = attr(object, 'weights'),
   replications = 200, ...) {
+    method = match.arg(method)
+    if(method == 'bootstrap') { 
+	se = bootstrap_se(object, replications)
+    } else if(method == 'jackknife') {
+	se = jackknife_se(object)
+    } else if(method == 'placebo') {
+	se = placebo_se(object, replications)
+    }
+    matrix(se^2)
+}
+
+#' Calculate the standard error of a synthetic diff in diff estimate. Deprecated. Use vcov.synthdid_estimate.
+#' @param ... Any valid arguments for vcov.synthdid_estimate
+#' @export synthdid_se
+synthdid_se = function(...) { sqrt(vcov(...)) }
+
+
+# The bootstrap se: Algorithm 2 of Arkhangelsky et al.
+bootstrap_se = function(estimate, replications) { sqrt((replications-1)/replications) * sd(bootstrap_sample(estimate, replications)) }
+bootstrap_sample = function(estimate, replications) { 
+    setup = attr(estimate, 'setup')
+    opts = attr(estimate, 'opts')
+    weights = attr(estimate, 'weights')
+    if (setup$N0 == nrow(setup$Y) - 1) { return(NA) }
+    theta = function(ind) {
+	if(all(ind <= setup$N0) || all(ind > setup$N0)) { NA }
+	else {
+	    weights.boot = weights
+	    weights.boot$omega = sum_normalize(weights$omega[sort(ind[ind <= setup$N0])])
+	    do.call(synthdid_estimate, c(list(Y=setup$Y[sort(ind),], N0=sum(ind <= setup$N0), T0=setup$T0, X=setup$X[sort(ind), ,], weights=weights.boot), opts))
+	}
+    }
+    bootstrap.estimates = rep(NA, replications)
+    count = 0
+    while(count < replications) {
+	bootstrap.estimates[count+1] = theta(sample(1:nrow(setup$Y), replace=TRUE))
+	if(!is.na(bootstrap.estimates[count+1])) { count = count+1 }
+    }
+    bootstrap.estimates
+}
+
+
+# The fixed-weights jackknife estimate of variance: Algorithm 3 of Arkhangelsky et al.
+# if weights = NULL is passed explicitly, calculates the usual jackknife estimate of variance.
+# returns NA if there is one treated unit or, for the fixed-weights jackknife, one control with nonzero weight
+jackknife_se = function(estimate, weights = attr(estimate, 'weights')) { 
+    setup = attr(estimate, 'setup')
+    opts = attr(estimate, 'opts')
+    if (setup$N0 == nrow(setup$Y) - 1 || (!is.null(weights) && sum(weights$omega != 0) == 1)) { return(NA) }
+    theta = function(ind) {
+	weights.jk = weights
+	if (!is.null(weights)) { weights.jk$omega = sum_normalize(weights$omega[ind[ind <= setup$N0]]) }
+	estimate.jk = do.call(synthdid_estimate, 
```

Comment:
there is a bug here, `opts` contains `update.omega = TRUE` and `update.lambda = TRUE`, causing the following line to recompute weights at every jackknife iteration